include/dxc/Support/WinAdapter.h

```diff
@@ -563,27 +566,65 @@ enum tagSTATFLAG {
 
 #ifdef __EMULATE_UUID
 
-size_t UuidStrHash(const char* k);
-
 // The following macros are defined to facilitate the lack of 'uuid' on Linux.
-#define DECLARE_CROSS_PLATFORM_UUIDOF(T)                                       \
-public:                                                                        \
-  static REFIID uuidof() { return reinterpret_cast<REFIID>(T##_ID); }          \
-                                                                               \
-private:                                                                       \
-  __attribute__((visibility("default"))) static const size_t T##_ID;
-
-#define DEFINE_CROSS_PLATFORM_UUIDOF(T)                                        \
-  __attribute__((visibility("default"))) const size_t T::T##_ID =              \
-      UuidStrHash(#T);
-#define __uuidof(T) T::uuidof()
+
+template <typename TInterface> struct TInterfaceIids {};
+
+#define DECLARE_CROSS_PLATFORM_UUIDOF_VALUE(interface_name, ...)               \
+  template <> struct TInterfaceIids<interface_name> {                          \
+    static REFIID uuidof() {                                                   \
+      static IID _IID = __VA_ARGS__;                                           \
+      return _IID;                                                             \
+    }                                                                          \
+  };
+
+// TODO: Move DECLARE_CROSS_PLATFORM_UUIDOF_VALUE into INTERFACE_STRUCT_HEADER
+
+#define INTERFACE_STRUCT_HEADER(interface_name, uuid0, uuid1, uuid2, uuid3_0,  \
+                                uuid3_1, uuid3_2, uuid3_3, uuid3_4, uuid3_5,   \
+                                uuid3_6, uuid3_7)                              \
+  struct interface_name;                                                       \
+  DECLARE_CROSS_PLATFORM_UUIDOF_VALUE(                                         \
+      interface_name, {0x##uuid0,                                              \
+                       0x##uuid1,                                              \
+                       0x##uuid2,                                              \
+                       {0x##uuid3_0, 0x##uuid3_1, 0x##uuid3_2, 0x##uuid3_3,    \
+                        0x##uuid3_4, 0x##uuid3_5, 0x##uuid3_6, 0x##uuid3_7}})  \
+  struct interface_name
+
+// TODO: We can also eat the open bracket and emit members containing the UUID.
+// Advantage:
+//    One can define subclasses on guid types, and __uuidof() still works
+//    normally on them (as TInterfaceIids won't have a specialization for that
+//    type - or is that impossible in Windows too?)
+// Disadvantage:
+//    public/private inheritance has to run through the struct
+//
+// This seems like a cakewalk tradeoff IF __uuidof() works on non-guid
+// subtypes in Windows
```

Comment:
I find these comments confusing. Can you elaborate?

Reply:
The tradeoffs are more clearly elaborated in the PR description and linked comment:

> > One major downside of this method, further elaborated in [#2680 (comment)](https://github.com/microsoft/DirectXShaderCompiler/issues/2680#issuecomment-635499459), is `__uuidof()` not working on an interface implementation (subclass). That might be solved by consuming the inheritance chain and `{` bracket in the macro and emitting the static `uuidof()` method again.